sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/WebExceptionRetryPolicy.java

```diff
@@ -33,37 +36,77 @@ public WebExceptionRetryPolicy() {
-        Duration backoffTime = Duration.ofSeconds(0);
-
-        if (!WebExceptionUtility.isWebExceptionRetriable(exception)) {
-            // Have caller propagate original exception.
+    public Mono<ShouldRetryResult> shouldRetry(Exception e) {
+        boolean isOutOfRetries = isOutOfRetries();
+        if (isOutOfRetries) {
             this.durationTimer.stop();
-            return Mono.just(ShouldRetryResult.noRetryOnNonRelatedException());
+            return Mono.just(ShouldRetryResult.noRetry());
         }
 
-        // Don't penalise first retry with delay.
-        if (attemptCount++ > 1) {
-            int remainingSeconds = WebExceptionRetryPolicy.waitTimeInSeconds - Math.toIntExact(this.durationTimer.getTime(TimeUnit.SECONDS));
-            if (remainingSeconds <= 0) {
-                this.durationTimer.stop();
-                return Mono.just(ShouldRetryResult.noRetry());
+        // Received Connection error (HttpRequestException), initiate the endpoint rediscovery
+        CosmosException webException = Utils.as(e, CosmosException.class);
+        if (WebExceptionUtility.isNetworkFailure(e) && this.isReadRequest &&
+            (webException != null && WebExceptionUtility.isReadTimeoutException(webException) &&
+                Exceptions.isSubStatusCode(webException, HttpConstants.SubStatusCodes.GATEWAY_ENDPOINT_READ_TIMEOUT))) {
+
+            // if operationType AddressRefresh then just retry
+            if (this.request.isAddressRefresh()) {
+                return shouldRetryAddressRefresh();
             }
 
-            backoffTime = Duration.ofSeconds(Math.min(this.currentBackoffSeconds, remainingSeconds));
-            this.currentBackoffSeconds *= WebExceptionRetryPolicy.backoffMultiplier;
+            return Mono.just(ShouldRetryResult.retryAfter(Duration.ofSeconds(retryDelay)));
         }
 
-        logger.warn("Received retriable web exception, will retry", exception);
+        if (!WebExceptionUtility.isWebExceptionRetriable(e)) {
+            // Have caller propagate original exception.
+            this.durationTimer.stop();
+            return Mono.just(ShouldRetryResult.noRetryOnNonRelatedException());
+        }
 
-        return Mono.just(ShouldRetryResult.retryAfter(backoffTime));
+        logger.warn("Received retriable web exception, will retry", e);
+        return Mono.just(ShouldRetryResult.retryAfter(Duration.ofSeconds(retryDelay)));
     }
 
     @Override
     public RetryContext getRetryContext() {
         return this.retryContext;
     }
+
+    public void onBeforeSendRequest(RxDocumentServiceRequest request) {
+        this.request = request;
+        this.isReadRequest = request.isReadOnlyRequest();
+        this.timeoutPolicy = HttpTimeoutPolicy.getTimeoutPolicy(request);
+        // Fetching the retryCount to correctly get the retry values from the timeout policy
+        if (this.retryContext != null) {
+            this.retryCountTimeout = this.retryContext.getRetryCount();
```

Comment:
There is no increase of `retryCountTimeout`?

Reply:
I am making use of the retryContext.retryCount to increase the retryCountTimeout in the code.